llvm/lib/ProfileData/InstrProfWriter.cpp

```diff
@@ -485,6 +489,42 @@ static uint64_t writeMemProfFrames(
   return FrameTableGenerator.Emit(OS.OS);
 }
 
+// Serialize MemProfFrameData.  Return the mapping from FrameIds to their
+// indexes within the frame array.
+static llvm::DenseMap<memprof::FrameId, uint32_t> writeMemProfFrameArray(
+    ProfOStream &OS,
+    llvm::MapVector<memprof::FrameId, memprof::Frame> &MemProfFrameData) {
+  // Mappings from FrameIds to array indexes.
+  llvm::DenseMap<memprof::FrameId, uint32_t> MemProfFrameIndexes;
+
+  // Sort the FrameIDs for stability.
+  std::vector<memprof::FrameId> FrameIdOrder;
+  FrameIdOrder.reserve(MemProfFrameData.size());
+  for (const auto &KV : MemProfFrameData)
+    FrameIdOrder.push_back(KV.first);
+  assert(MemProfFrameData.size() == FrameIdOrder.size());
+  llvm::sort(FrameIdOrder);
+
+  // Serialize all frames while creating mappings from linear IDs to FrameIds.
+  uint64_t Index = 0;
+  MemProfFrameIndexes.reserve(FrameIdOrder.size());
+  for (memprof::FrameId Id : FrameIdOrder) {
+    auto Iter = MemProfFrameData.find(Id);
```

Comment:
If this loop has non-trivial time overhead, another possibility would be to convert FrameIdOrder to a vector of pairs (of the frame id and frame), then sort it based on the frame id. That avoids an extra loop here of map lookups.

Reply:
Good point!  Fixed in the latest iteration.